Fix condition in JACK backend check, previously resulting in too large buffers

robbert-vdh opened the pull request:
`m_deviceInfo->hostApi` is an index in the list of supported APIs on this platform. So on Linux 0 is ALSA, 2 is JACK. This is also mentioned in the comment on the `hostApi` field. We should be checking the device type ID instead, as is done elsewhere in the code.

This fixes the buffer size when using JACK clients. With JACK the buffer size should be coming from the server (and PortAudio chops it up into 1024 sample blocks if it's larger). Before this change it would use the now grayed out latency value set for the previous audio backend before switching to JACK (defaulting to 1024 samples).

Setting `m_framesPerBuffer` to 0 causes causes PortAudio to use the JACK server's period size like you'd expect from a client. `m_framesPerBuffer` is also used to size some FIFOs but I _think_ those are never used with JACK (and before this change those would have also had an incorrect size anyways). I'd appreciate it if someone who knows more about Mixxx's audio backend could confirm this though.

robbert-vdh committed:
Fix condition in JACK backend check

`m_deviceInfo->hostApi` is an index in the list of supported APIs on
this platform. So on Linux 0 is ALSA, 2 is JACK. This is also mentioned
in the comment on the `hostApi` field. We should be checking the device
type ID instead, as is done elsewhere in the code.

This fixes the buffer size when using JACK clients. With JACK the buffer
size should be coming from the server (and PortAudio chops it up into
1024 sample blocks if it's larger). Before this change it would use the
now grayed out latency value set for the previous audio backend before
switching to JACK (defaulting to 1024 samples).

daschuer commented:
Cool, great that you have found this issue. Since the stable 2.3 branch is also affected, can you rebase it to that branch? 
```
git rebase --onto=upstream/2.3 HEAD~1 
git push -f
```

In addition we need you permission to distribute your changes with Mixxx. Please sign: 
https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
And comment here when done. 
Thank you.

robbert-vdh commented:
Yes of course, wasn't sure whether to target main or 2.3 for this fix since it technically does affect behavior (though it should only improve things). Rebased onto 2.3 (does this then get cherry picked back into master?), and signed the CLA.

ronso0 commented:
After changes to 2.3 that is merged into main, so no worries, your change will land in main soonish.

robbert-vdh commented:
I also confirmed that none of the code paths that would do the wrong thing when the `m_framesPerBuffer == 0` are hit when using the JACK backend. Still feels like this could probably be abstracted away in a safer manner.

daschuer commented:
It looks like the code never works as desired. For my understanding, the check for paJack needs to be moved to: 
SoundDevice::setFramesPerBuffer 

Portaudio uses a fixed temp buffer of 1024 in case of Jack: 
https://github.com/PortAudio/portaudio/blob/06bc8763e72b6aa5fa1f380946e09b676e09806f/src/common/pa_process.c#L54
I have pipewire installed and I see the issue that this defines the maximum buffer. 
When I start Mixxx with  
`PIPEWIRE_LATENCY=2048/48000 ./mixxx`
It performs two callbacks in a row. This makes the smoothing topic even more difficult.  

With this PR I see also the Waveform jerking a lot. This is because the waveform smoothing code around VisualPlayPosition and VsyncThread is using the wrong buffer size. 

The Portaudio code is written in a way the the buffer can change at any time. 
PaUtil_SetInputFrameCount() is called every callback. 

This means we need to do the same and use the value passed to us via 
SoundDevicePortAudio::callbackProcessClkRef() 

There seems also an issue when trying to run Mixxx with 44100 Hz. 
`Warning [Main]: Error opening stream: Invalid sample rate`

Unfortunately this puts many places in Mixxx to a challenge.

daschuer reviewed (changes requested):
Conclusion, this PR alone does not work. 
I suggest to issue a small band aid PR the forces 1024/48000 and postpone the dynamic buffer size fix to a second PR.

robbert-vdh commented:
Did you try specifying smaller than 1024 quantum sizes with PipeWire? This worked as expected for both me and @Be-ing. I ran `pw-metadata -n settings 0 clock.force-quantum 128` before opening Mixxx, which causes the process function to be called with 128 sample buffers. Lots of software chop larger buffers up into 1024 sample blocks since DSP that doesn't already use block splitting may perform very bad at larger sizes due to the bad cache locality. PortAudio may be doign the same thing.

ronso0 commented:
https://mixxx.zulipchat.com/#narrow/stream/247620-development-help/topic/Incorrect.20Jack.20buffer.20size

daschuer commented:
> Did you try specifying smaller than 1024 quantum sizes with PipeWire? 

Yes, this was my first attempt leading the jerking wafeforms. 

There is lot of code using m_framesPerBuffer. For the final solution all this needs to made aware of a changing buffer size. It looks like we should get rid of this so member entirely and use always, the buffer size of the callback. 

This feels like too risky for our stable branch, that's why I have proposed to do a fixed 1024 solution that at least fixes the dependency to the grayed out buffer size.

Be-ing commented:
Hooray for C's implicit conversions. :roll_eyes: 

The current situation is that JACK doesn't actually go to low latencies without a very unintuitive workaround of configuring the buffer size with ALSA first, then switching to JACK. The current behavior is plainly incorrect and IMO should be fixed urgently! I can reproduce the waveform jerking with buffer sizes > 1024. That should be fixed too, but I think it's beyond the scope of this one line bugfix for the stable branch. PipeWire's default buffer size is 1024 and users are unlikely to be manually configuring either PipeWire or JACK to buffer sizes > 1024 to use Mixxx, so I think it's acceptable to leave the waveform jerking bug unfixed for now to fix this bug. I think this should be merged now as it is.

Be-ing commented:
As a temporary stopgap, the user could be warned when using the JACK PortAudio backend with > 1024 frames per buffer.

daschuer commented:
> I think this PR should be merged now as it is.

No, this does not suit for a stable branch. With this PR merged, some parts of Mixxx are using the actual buffer size, some zero and some the original preferences option. In addition the actual buffers size can change at any time. This state is IMHO worse compared to the current state where all values have the same value, but taken from a grayed out preferences option.

robbert-vdh commented:
Where else is `m_framesPerBuffer` used? I did a cursory search but couldn't find anything that would be applicable when using the JACK backend. And the grayed out preference value? 

If `m_framesPerBuffer` is still used in places, then another (safer, though still not great) option would be to set `m_framesPerBuffer` to `PA_FRAMES_PER_TEMP_BUFFER_WHEN_HOST_BUFFER_SIZE_IS_UNKNOWN_` instead of setting it to `paFramesPerBufferUnspecified`. The `paFramesPerBufferUnspecified` can then be passed directly to `Pa_OpenStream()`. That avoids avoid the 0 value from causing issues. As long as all other code uses the actual buffer size passed to the audio callback when interacting with data structures that should work great.

daschuer commented:
I have just skimmed over the code and revealed this: 
 
https://github.com/mixxxdj/mixxx/blob/e28e0411063f519c7fb55f25d9d898d241ba6d99/src/soundio/soundmanager.cpp#L694

One relevant point is here: 
https://github.com/mixxxdj/mixxx/blob/main/src/soundio/sounddeviceportaudio.cpp#L1017

There are some other places in SoundDevicePortAudio/SoundDevice that are probably not really used, but at a minimum a safety net is missing. This is related to Multi sound card mode which is anyway broken with Jack and Pipewire and Network clock mode.

Not sure if there is more. If that is really all a proper fix seems less invasive than originally thought. 

Setting m_framesPerBuffer to PA_FRAMES_PER_TEMP_BUFFER_WHEN_HOST_BUFFER_SIZE_IS_UNKNOWN_ would be a reliable quick fix that always works. Portaudio takes care that the framesPerBuffer does not change.  

Another idea would be just to un-gray-out the preferences option ...

Be-ing commented:
> Another idea would be just to un-gray-out the preferences option ...

JACK allows clients to set the buffer size with [jack_set_buffer_size](https://jackaudio.org/api/group__ServerControl.html#gad96bda0901b924f5996e02e41a979685). Bizarrely PortAudio doesn't call that function though.

Be-ing commented:
Another option would be to set m_framesPerBuffer from the PaStreamInfo after calling Pa_StartStream, though I don't know what effects that might have with other PortAudio backends.

Be-ing commented:
> JACK allows clients to set the buffer size with [jack_set_buffer_size](https://jackaudio.org/api/group__ServerControl.html#gad96bda0901b924f5996e02e41a979685). Bizarrely PortAudio doesn't call that function though.

I forgot I already [reported that upstream](https://github.com/PortAudio/portaudio/issues/507) a while ago... and this seems to be by design :weary: I am questioning whether squeezing JACK into PortAudio's design is really a good idea.

robbert-vdh commented:
> I have just skimmed over the code and revealed this:
> 
> https://github.com/mixxxdj/mixxx/blob/e28e0411063f519c7fb55f25d9d898d241ba6d99/src/soundio/soundmanager.cpp#L694

Do you happen to know those things use `getFramesPerBuffer()`/`m_framesPerBuffer` instead of the number of samples passed to the audio callback? That would make way more sense to me. There's now  also the duplication/functional dependency between `m_framesPerBuffer` and `m_audioBufferSizeIndex`, which I guess this PR breaks by fixing the conditional that changes the former.

> > Another idea would be just to un-gray-out the preferences option ...
> 
> JACK allows clients to set the buffer size with [jack_set_buffer_size](https://jackaudio.org/api/group__ServerControl.html#gad96bda0901b924f5996e02e41a979685). Bizarrely PortAudio doesn't call that function though.

That would open up another can of worms. I don't think Mixxx should be configuring the server. It should just use whatever the JACK server is currently using. But yeah PortAudio's JACK backend seems pretty suboptimal, and I've heard bad things about it over the years.

Be-ing commented:
I was concerned PortAudio's buffer size adaptation might be adding latency unconditionally with JACK, but on closer reading of the PortAudio code, that might happen only if the framesPerBuffer is specified to Pa_OpenStream. That is, PortAudio may add latency if the framesPerBuffer passed to Pa_OpenStream is different from the JACK buffer size, which is currently the case because framesPerBuffer comes from whatever it happened to be last set to when using the ALSA backend. Mixxx can't avoid this by passing the JACK server's buffer size to Pa_OpenStream because PortAudio doesn't expose an API to query that information from the JACK server before opening the stream. Regardless of the JACK server's buffer size, PortAudio will still call the application's callback with the framesPerBuffer passed to Pa_OpenStream, but under the hood the buffer size adaptation mechanism could be adding latency in the likely case where the framesPerBuffer specified to Pa_OpenStream does not match the JACK server buffer size. :grimacing: 

When Pa_OpenStream is passed paFramesPerBufferUnspecified (0) for framesPerBuffer, PortAudio [uses its non-adapting process](https://github.com/PortAudio/portaudio/blob/06bc8763e72b6aa5fa1f380946e09b676e09806f/src/common/pa_process.c#L134). NonAdaptingProcess uses a do-while loop that [clamps the framesPerBuffer passed to the application callback](https://github.com/PortAudio/portaudio/blob/06bc8763e72b6aa5fa1f380946e09b676e09806f/src/common/pa_process.c#L734) to the PA_FRAMES_PER_TEMP_BUFFER_WHEN_HOST_BUFFER_SIZE_IS_UNKNOWN_ constant (1024) and calls the application's callback multiple times in case the host (JACK server) buffer size is greater than that arbitrary 1024 constant. In case the JACK server buffer size <= 1024, the application callback gets called once per JACK callback and no latency is added.

So, the original intent of the broken code that this PR is fixing is correct. Pa_OpenStream should be passed paFramesPerBufferUnspecified when using JACK to avoid adding unnecessary latency. In the case that framesPerBuffer is specified to Pa_OpenStream, IMO it is surprising that PortAudio may add latency behind the application's back rather than tell the JACK server to change its buffer size to match the requested buffer size. However, JACK users generally don't expect JACK applications besides control applications like QJackCtl or Catia to set the JACK buffer size. So I'm not going to bother working upstream to change what PortAudio does when framesPerBuffer is specified with the JACK backend. It's not nice that these details are not documented and I had to dig into PortAudio's implementation details to figure this out. :confused:

daschuer commented:
> That would make way more sense to me.

To me as well. The function should receive the actual frames per buffer as parameter. 

I also consider the SoundDevice::m_framesPerBuffer badly named, because in case of SoundDevicePortaudio, it is the configuredFramesPerBuffer or such, not the actual. Maybe we should rename this. 

Do you have interest to issue another PR with these changes? This seems to be a low risk change ans will help to have a clearer view on this here. 

> That would open up another can of worms.

Fortunately not. It configures only the Portaudio client buffer that is passed to us. Portaudio hides the actual Server buffer size from us and keeps our buffer size const even if the server buffer changes. 

> But yeah PortAudio's JACK backend seems pretty suboptimal

Yes, we have some open JACK bugs, the main issue is: 
https://github.com/mixxxdj/mixxx/issues/6998

A while a go I have started to implement a SoundDeviceJack to make Mixxx a native Jack Application exposing all outputs as Jack ports. Unfortunately the work has stalled: https://github.com/daschuer/mixxx/tree/jacknative

Be-ing commented:
I agree that consistently using the framesPerBuffer passed from PortAudio to paV19Callback instead of the configured value would make sense.

Be-ing commented:
FWIW it looks like this code may have worked in the past, but was broken by f2942e2c39178f6b4b84c27b24bd0f62ba45f5ef.

Swiftb0y commented:
fyi because issues like this and other concerning PortAudio/PortMidi (no hotplugging for example), I'd propose to replace the `Port*` libraries with their `Rt*` equivalences...

daschuer commented:
@robbert-vdh I consider add this PR to our upcoming 2.3.4. release? Will it be OK, when I will finish this PR this weekend?

robbert-vdh commented:
@daschuer Yes, of course! I wanted to also change the uses of `m_framesPerBuffer`/`m_config.getFramesPerBuffer()` in the DSP code to use the provided buffer size instead but I haven't gotten around to it yet (and I won't be getting around to it for at least a couple more days).

robbert-vdh committed:
Replace m_framesPerBuffer uses with runtime size

In any operation that works with audio buffers passed from PortAudio.
This is needed because PortAudio may pass us buffers that are smaller or
larger than the size we request. And when using JACK the requested size
is 0, which will cause PortAudio to pick a buffer size matching the JACK
server.

robbert-vdh committed:
Use std::unique_ptr for FIFOs in PortAudio backend

The network backend already does this.

robbert-vdh committed:
Expose the current FIFO size

robbert-vdh committed:
Explicitly check when FIFOs are too small

This situation should never happen, but in case it does come up we
should probably decide on what to do next. Resizing the FIFOs on the
audio thread may result in a momentary audio glitch or xrun but after
that everything would work fine again.

robbert-vdh committed:
Remove unused getProcessingLatency() function

robbert-vdh committed:
Broadcast buffer size changes to the GUI

This fixes the waveforms moving at the wrong pace and subsequently
snapping back to the correct position at every process call as mentioned
in https://github.com/mixxxdj/mixxx/pull/11092#issuecomment-1326879866.

robbert-vdh commented:
Okay, I changed the (PortAudio) sound device code to use the provided buffer size, and I made sure buffer size changes get broadcasted to the GUI. This fixes the waveform jerking back and forth and runtime buffer changes with PipeWire are also handled correctly. I'll go through everything one more time and then submit another PR

Be-ing commented:
You can push to this branch instead of scattering the discussion with a new PR.

daschuer commented:
Oh, that's unfortunate. I'm working on a solution right now, too:
https://github.com/daschuer/mixxx/tree/jack-api-check
Hopefully we can merge them to one finally. 
Currently I am fighting with a penalty in WaveformRendererPreroll::draw() discovered during debugging.

daschuer commented:
> If you want I can cherry pick the changes from https://github.com/mixxxdj/mixxx/pull/11114 into this branch and close that other PR.

Yes that would be better. You can even reset this branch and than force push as [jack-api-check](https://github.com/robbert-vdh/mixxx/tree/fix/jack-api-check)

robbert-vdh commented:
I rebased onto the current 2.3 branch and added the commits from #11114. For completeness' sake, this was the PR's description:

<details>
<summary>PR #11114</summary>

This is a followup to #11092, which allowed Mixxx to use the JACK server's configured period size instead of still using the value from the grayed out latency field in the settings. The old behavior potentially introduced unnecessary buffering and made controls less responsive. @daschuer noticed that when using a very large buffer sizes in combination with a small configured latency the waveform display was jerking all over the place. As part of fixing that, this PR gets rid of all uses of `m_framesPerBuffer` and values derived from that in the PortAudio process calls and all functions called from there. The network streaming backend always uses the same size so while it has been changed similarly for consistency's sake, that still ends up using `m_framesPerBuffer` for the buffer size. The waveform display and other GUI elements that rely on the buffer size now also correctly respond to buffer size changes and thus work as expected in combination with the change from #11092 regardless of the JACK server's settings.

One part where I wasn't sure what to do was with the input and output FIFOs in the PortAudio backend. Those do not handle runtime buffer sizes that are larger than the configured sizes. They are not used for JACK and if this has never caused issues with earlier Mixxx versions, then that means that the buffer size will always be equal to `m_framesPerBuffer`, but I still added another very visible warning just in case. Another option would be to adapt to new buffer size and just reallocate the FIFOs right there and then. That would at worst introduce an xrun when starting Mixxx, but I have not yet done so since this situation doesn't seem to come up in practice.

I also targeted the 2.3 branch to be in line with the other PR. Let me know if you want me to rebase either or both PRs onto main instead.

</details>

And my other comment on that PR:

> Oh and I also did not rename `m_framesPerBuffer` and `getFramesPerBuffer()` yet. Their names should more clearly indicate that they only return reflect the configured size, and that the actual runtime size may be different. I really wanted to, but that felt out of scope for this PR.

daschuer commented on src/soundio/sounddevice.h:
```diff
@@ -84,9 +91,21 @@ class SoundDevice {
+    /**
+     * The last buffer size that was sent to the `[Master],audio_buffer_size`
+     * control object. This is set in `maybeUpdateBufferSize()`.
+     */
```
We use C style comments for disable code. Please prefer // for comments. /// is picked up by doxygen if you like.

daschuer replied:
Yes, see:
https://github.com/mixxxdj/mixxx/wiki/Coding-Guidelines#comments

daschuer commented:
Thank you for this updated PR. I will incorporate your commits into my branch. 
I am currently working to get a multi sound-card setup to work under al circumstances.

robbert-vdh committed:
Use C#/Rust style doc comments instead of javadoc

As requested.

daschuer commented:
@robbert-vdh I have rebased my work on top of yours. https://github.com/mixxxdj/mixxx/pull/11121
I have extended it to not use a stored buffer size at all. Two soundcards are also working with changing buffers.

robbert-vdh commented:
> @robbert-vdh I have rebased my work on top of yours. #11121 I have extended it to not use a stored buffer size at all. Two soundcards are also working with changing buffers.

Awesome!

daschuer commented:
Can we close this one now in favour of https://github.com/mixxxdj/mixxx/pull/11121

robbert-vdh commented:
Yes this can be closed when that PR gets merged.

robbert-vdh commented:
Can be closed now that #11121 is merged.
